Use smart pointers with WebCore::CachedResourceLoader in libxml2/libxslt code

ddkilzer opened the pull request:
#### 49e16e9ddcee72f8f19370735909ca1a677a8b14
<pre>
Use smart pointers with WebCore::CachedResourceLoader in libxml2/libxslt code
<a href="https://bugs.webkit.org/show_bug.cgi?id=273204">https://bugs.webkit.org/show_bug.cgi?id=273204</a>
&lt;<a href="https://rdar.apple.com/126997686">rdar://126997686</a>&gt;

Reviewed by Chris Dumez and Alex Christensen.

Deploy smart pointers for CachedResourceLoader objects used in
libxml2/libxslt integration code.  Use WeakPtr&lt;&gt; for storing temporary
references.  Use RefPtr&lt;&gt; when used multiple times in a method.

Also change 0 to nullptr functions that were modified.

* Source/WebCore/xml/XSLTProcessorLibxslt.cpp:
(WebCore::globalCachedResourceLoader): Add.
(WebCore::docLoaderFunc):
- Add early return when there is no data to parse.  This also simplifies
  the arguments to xmlReadMemory().
(WebCore::setXSLTLoadCallBack):
(WebCore::XSLTProcessor::transformToString):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::matchFunc):
(WebCore::shouldAllowExternalLoad):
(WebCore::openFunc):
- Add early return when
  XMLDocumentParserScope::currentCachedResourceLoader() is nullptr.
- Store Document in a RefPtr.
(WebCore::XMLDocumentParser::doEnd):
* Source/WebCore/xml/parser/XMLDocumentParserScope.cpp:
(WebCore::XMLDocumentParserScope::currentCachedResourceLoader): Add.
(WebCore::XMLDocumentParserScope::XMLDocumentParserScope):
(WebCore::XMLDocumentParserScope::~XMLDocumentParserScope):
* Source/WebCore/xml/parser/XMLDocumentParserScope.h:
- Fix old-style indentation of class declaration.
- Delete unused default constructor.
- Provide default initializers for instance variables.

Canonical link: <a href="https://commits.webkit.org/277956@main">https://commits.webkit.org/277956@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/02d6e013b5adc87f17307f95fe35149338b60cc5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49042 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28260 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52011 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51760 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/45113 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51346 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34236 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25793 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40082 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/50293 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25906 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42272 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21190 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23354 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43442 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7156 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/45285 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43947 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53668 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24104 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20345 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47397 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/25385 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42480 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46372 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26172 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7026 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/25106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/xml/XSLStyleSheetLibxslt.cpp:
```diff
@@ -117,12 +117,12 @@ void XSLStyleSheet::clearXSLStylesheetDocument()
     if (!document)
         return nullptr;
-    return &document->cachedResourceLoader();
+    return document->protectedCachedResourceLoader();
```
This change is not needed. It is more verbose and has not benefit. Per the guidelines, `protectedFoo()` functions are only useful to avoid local variables.

ddkilzer replied:
FIXED

cdumez commented on Source/WebCore/xml/XSLStyleSheetLibxslt.cpp:
```diff
@@ -117,12 +117,12 @@ void XSLStyleSheet::clearXSLStylesheetDocument()
 }
 
-CachedResourceLoader* XSLStyleSheet::cachedResourceLoader()
+RefPtr<CachedResourceLoader> XSLStyleSheet::cachedResourceLoader()
```
There is no ownership transfer here, we normally keep returning a raw pointer in such cases to avoid unnecessary ref counting churn for call sites that don't need it.

ddkilzer replied:
FIXED

cdumez commented on Source/WebCore/xml/XSLStyleSheet.h:
```diff
@@ -71,7 +71,7 @@ class XSLStyleSheet final : public StyleSheet, public CanMakeWeakPtr<XSLStyleShe
     void loadChildSheet(const String& href);
 
-    CachedResourceLoader* cachedResourceLoader();
+    RefPtr<CachedResourceLoader> cachedResourceLoader();
```
Doesn't match our smart pointer guidelines.

cdumez replied:
If we needed a function that returned RefPtr (I didn't check if this was truly needed in this PR), it would be called `protectedCachedResourceLoader()`.

ddkilzer replied:
FIXED

cdumez reviewed (changes requested):
Doesn't match the smart pointer guidelines.

cdumez commented on Source/WebCore/xml/XSLTProcessorLibxslt.cpp:
```diff
@@ -146,14 +156,14 @@ static xmlDocPtr docLoaderFunc(const xmlChar* uri,
 }
 
-static inline void setXSLTLoadCallBack(xsltDocLoaderFunc func, XSLTProcessor* processor, CachedResourceLoader* cachedResourceLoader)
+static inline void setXSLTLoadCallBack(xsltDocLoaderFunc func, XSLTProcessor* processor, WeakPtr<CachedResourceLoader> cachedResourceLoader)
```
We don't need to change the parameter type here. Nothing in our guidelines indicates to update parameter types.

ddkilzer replied:
FIXED

cdumez commented on Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
```diff
@@ -466,11 +467,14 @@ static bool shouldAllowExternalLoad(const URL& url)
+    if (!cachedResourceLoader)
+        return &globalDescriptor;
+
+    Document* document = cachedResourceLoader->document();
```
Why aren't we using a RefPtr for the document here?

ddkilzer replied:
Because the patch was to fix `CachedResourceLoader`.

The scope already expected quite a bit just doing that.  If I added `Document`, the patch would have gotten even larger.

ddkilzer replied:
I'll make this change for this function.

cdumez replied:
I'm also fine with you sticking to CachedResourceLoader in this patch.

ddkilzer replied:
Since I was already modifying this line, I might as well make the change.  :)

cdumez commented on Source/WebCore/xml/parser/XMLDocumentParserScope.cpp:
```diff
@@ -28,27 +28,33 @@
-    , m_oldStructuredErrorFunc(xmlStructuredError)
-    , m_oldErrorContext(xmlGenericErrorContext)
-#endif
+XMLDocumentParserScope::XMLDocumentParserScope(WeakPtr<CachedResourceLoader> cachedResourceLoader)
```
We shouldn't pass a WeakPtr in parameter here. The previous code matches the guidelines.

ddkilzer replied:
FIXED

cdumez commented on Source/WebCore/xml/parser/XMLDocumentParserScope.cpp:
```diff
@@ -28,27 +28,33 @@
+{
+}
+#else
+XMLDocumentParserScope::XMLDocumentParserScope(WeakPtr<CachedResourceLoader> cachedResourceLoader)
```
ditto.

ddkilzer replied:
FIXED

cdumez commented on Source/WebCore/xml/parser/XMLDocumentParserScope.cpp:
```diff
@@ -28,27 +28,33 @@
 #if ENABLE(XSLT)
-XMLDocumentParserScope::XMLDocumentParserScope(CachedResourceLoader* cachedResourceLoader, xmlGenericErrorFunc genericErrorFunc, xmlStructuredErrorFunc structuredErrorFunc, void* errorContext)
-    : m_oldCachedResourceLoader(currentCachedResourceLoader)
+XMLDocumentParserScope::XMLDocumentParserScope(WeakPtr<CachedResourceLoader> cachedResourceLoader, xmlGenericErrorFunc genericErrorFunc, xmlStructuredErrorFunc structuredErrorFunc, void* errorContext)
```
ditto.

ddkilzer replied:
FIXED

cdumez commented on Source/WebCore/xml/parser/XMLDocumentParserScope.h:
```diff
@@ -33,28 +33,29 @@
+    WTF_MAKE_NONCOPYABLE(XMLDocumentParserScope);
+public:
+    XMLDocumentParserScope() = delete;
+    explicit XMLDocumentParserScope(WeakPtr<CachedResourceLoader>);
```
ditto.

ddkilzer replied:
FIXED

cdumez commented on Source/WebCore/xml/parser/XMLDocumentParserScope.h:
```diff
@@ -33,28 +33,29 @@
 
 #if ENABLE(XSLT)
-        XMLDocumentParserScope(CachedResourceLoader*, xmlGenericErrorFunc, xmlStructuredErrorFunc structuredErrorFunc = 0, void* errorContext = nullptr);
+    XMLDocumentParserScope(WeakPtr<CachedResourceLoader>, xmlGenericErrorFunc, xmlStructuredErrorFunc = nullptr, void* errorContext = nullptr);
```
ditto.

ddkilzer replied:
FIXED

cdumez commented on Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
```diff
@@ -466,11 +467,14 @@ static bool shouldAllowExternalLoad(const URL& url)
+    if (!cachedResourceLoader)
+        return &globalDescriptor;
+
+    RefPtr document = cachedResourceLoader->protectedDocument();
```
`RefPtr document = cachedResourceLoader->document();`

ddkilzer replied:
I'm really curious why `document()` is preferred over `protectedDocument()` here.

I guess I don't understand the purpose of `protectedDocument()` if it's not used here.  Is it only supposed to be used internal to the class itself?

ddkilzer replied:
Ah, I see this is in the code guidelines.  The `protectedFoo()` functions are to avoid a local variable.

cdumez replied:
Calling `protectedDocument()` here doesn't hurt but it also has no benefit over calling `document()` since you're storing it in a RefPtr anyway. No benefit, not even perf. However, it is more verbose so there is no reason to change existing code.

`protectedDocument()` is indeed useful to avoid a local variable, e.g.:
```
protectedDocument()->doSomething();
```

ddkilzer replied:
Oops, meant to push a fix to replace `protectedDocument()` before landing this.  Will fix in a moment.

ddkilzer replied:
Follow-up PR:  https://github.com/WebKit/WebKit/pull/27724

cdumez reviewed (approved):
LGTM % nit.

Commit:
Use smart pointers with WebCore::CachedResourceLoader in libxml2/libxslt code
https://bugs.webkit.org/show_bug.cgi?id=273204
<rdar://126997686>

Reviewed by Chris Dumez and Alex Christensen.

Deploy smart pointers for CachedResourceLoader objects used in
libxml2/libxslt integration code.  Use WeakPtr<> for storing temporary
references.  Use RefPtr<> when used multiple times in a method.

Also change 0 to nullptr functions that were modified.

* Source/WebCore/xml/XSLTProcessorLibxslt.cpp:
(WebCore::globalCachedResourceLoader): Add.
(WebCore::docLoaderFunc):
- Add early return when there is no data to parse.  This also simplifies
  the arguments to xmlReadMemory().
(WebCore::setXSLTLoadCallBack):
(WebCore::XSLTProcessor::transformToString):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::matchFunc):
(WebCore::shouldAllowExternalLoad):
(WebCore::openFunc):
- Add early return when
  XMLDocumentParserScope::currentCachedResourceLoader() is nullptr.
- Store Document in a RefPtr.
(WebCore::XMLDocumentParser::doEnd):
* Source/WebCore/xml/parser/XMLDocumentParserScope.cpp:
(WebCore::XMLDocumentParserScope::currentCachedResourceLoader): Add.
(WebCore::XMLDocumentParserScope::XMLDocumentParserScope):
(WebCore::XMLDocumentParserScope::~XMLDocumentParserScope):
* Source/WebCore/xml/parser/XMLDocumentParserScope.h:
- Fix old-style indentation of class declaration.
- Delete unused default constructor.
- Provide default initializers for instance variables.

Canonical link: https://commits.webkit.org/277956@main
